web/pgadmin/tools/sqleditor/static/js/components/dialogs/MacrosDialog.jsx

```diff
@@ -89,59 +115,87 @@ const useStyles = makeStyles((theme)=>({
   },
 }));
 
-function getChangedMacros(macrosData, changeData) {
+function getChangedMacros(userMacrosData, changeData) {
   /* For backend, added, removed is changed. Convert all added, removed to changed. */
   let changed = [];
+
   for (const m of (changeData.macro.changed || [])) {
     let newM = {...m};
     if('id' in m) {
-      /* if key changed, clear prev and add new */
-      changed.push({id: m.mid, name: null, sql: null});
-      let em = _.find(macrosData, (d)=>d.mid==m.mid);
-      newM = {name: em.name, sql: em.sql, ...m};
+      let em = _.find(userMacrosData, (d)=>d.id==m.id);
+      newM = {name: m.name ? (m.name) : em.name , sql: m.sql ? m.sql : em.sql, mid: m.mid ? m.mid : em.mid, ...m};
     } else {
       newM.id = m.mid;
     }
-    delete newM.mid;
     changed.push(newM);
   }
   for (const m of (changeData.macro.deleted || [])) {
     changed.push({id: m.id, name: null, sql: null});
   }
   for (const m of (changeData.macro.added || [])) {
+    if (m.id && m.id !== 0){
+      m.mid = m.id;
+      delete m.id;
+    }
     changed.push(m);
   }
   return changed;
 }
 
+function getRandomName(existingDataCount, existingNames) {
+  let newName = `Macro ${existingDataCount + 1}`;
+  // Check if the new name already exists then generating new name.
+  while (existingNames.includes(newName)) {
+    existingDataCount++;
+    newName = `Macro ${existingDataCount + 1}`;
+  }
+  return newName;
+}
+
```

Comment:
A similar function is defined in the QueryToolComponent.jsx. Please make a common function and use it.